docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md

```diff
@@ -993,6 +994,82 @@ The `-fexceptions` option is also needed for C code that needs to interoperate w
+
+- Call verification: Performance penalty between 5% and 10%.
+- Object permission changes: Between 400% and 700% slowdown for permission changes per object file, while a performance loss of 320 ms is noticeable for permission changes per binary.
+- Virtual function hashtable size: Storage of virtual function hashtable imply a big waste of space.
+
+Benchmarks were produced using tools from the `SPEC CPU 2006 C++ benchmark` suite[^SPECCPU2006] (obsolete and replaced by SPEC CPU 2017 as of January 2018 [^SPECCPU2017]). `omnetpp`, `astar` and `xalancbmk` showed a performance penalty of between 2.4% and 19.6%. Four other benchmarks (`povray`, `namd`, `soplex` and `dealII`) had no significant effect on performance [^Tice2014].
+
+Compile-time optimization such as PGO, devirtualization and statically linked binaries to `libvtv` offer performance advantages. For example, the performance penalty on `xalancbmk` was reduced from 19.6% to 8.6% using these three optimizations[^Tice2014].
+
+This information is presented in graphs in two presentation[Tice2012slide][^Tice2014slide].
+
+Moreover, GCC's virtual table checking function is tracked by the `-fvtv-counts` flag[^gccvtvcount], which provides counters to evaluate the number of virtual calls checked and the size of virtual table pointer sets for each class. These counters can provide information for assessing the performance penalty on source code. This information is written to two log files:
+
+- `vtv_count_data.log` - The number of virtual calls being verified for each class.
+- `vtv_class_set_sizes.log` - The size of the vtable pointer sets for each class.
+
+[^Tice2012slide]: Tice, Caroline, [Improving Function Pointer Security for Virtual Method Dispatches (slide)](https://gcc.gnu.org/wiki/cauldron2012?action=AttachFile&do=get&target=cmtice.pdf#page=38), GNU Tools Cauldron, July 2012
+[^Tice2014slide]: Tice, Caroline, [Enforcing Forward-Edge Control-Flow Integrity in GCC & LLVM (slide)](https://www.usenix.org/sites/default/files/conference/protected-files/sec14_slides_tice.pdf#page=19), USENIX Security, August 2014
+[^Tice2014]: Tice, Caroline, [Enforcing Forward-Edge Control-Flow Integrity in GCC & LLVM](https://www.usenix.org/system/files/conference/usenixsecurity14/sec14-paper-tice.pdf#page=12) USENIX Security, August 2014
+[^gccvtvcount]: Team GCC, [Program Instrumentation Options: `-fvtv-counts`](https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html#index-fvtv-counts), GCC Manual, August 2013
+[^SPECCPU2006]: Standard Performance Evaluation Corporation, [SPEC CPU 2006 benchmack](https://www.spec.org/cpu2006/) SPEC, August 2006
+[^SPECCPU2017]: Standard Performance Evaluation Corporation, [SPEC CPU 2017 benchmack](https://www.spec.org/cpu2017/) SPEC, June 2017
+
+#### When not to use?
+
+Using the `-fvtable-verify` flag can lead to ABI breaks.
+Software that needs to interoperate with libraries built without -fvtable-verify should not enable this option.
```

Comment:
This includes the C++ standard library, and all system libraries. Those are typically not built with VTV enabled, so this option is only useful if you build your own GCC, and then self-build your entire software stack from the C++ std::lib upwards. In practice, that means it's useless.

I agree with Sam, this is just not good advice.

Reply:
I agree with the difficulty to maintain binaries at Linux systems level. This agree this flag implying a lot of complication in case of system libraries (linking issues, etc...). 

First let me situate this flag in this guide. The description of this guide is about produce guide lines to build application with security mechanisms not only building a whole OS.

My proposal is mainly based on my experience in HPC environments where it is acceptable to offer different versions of compilers, HPC applications and related libraries on the same userspace (user context or application context).

HPC applications can be compiled and installed to accept custom library configurations using environment variables such as LD_PRELOAD or LD_LIBRARY_PATH, and PATH. This type of userspace can be maintained using tools such as "environment-modules" (https://modules.readthedocs.io/en/latest/), where application binaries and related libraries are built outside the system tree, in a hierarchy of custom folders.  Custom hierarchies are commonly used to offer applications with custom compilation configuration flags.  The use of a customized folder hierarchy offering the application and associated libraries compiled with VTV becomes reliable.

Let me expose a real-life use-case where the -fvtable-verify flag can be acceptable:
- "environment-modules" allow to provide custom build of GCC with –enable-vtable-verify flag is acceptable in a custom folder hierarchy without influencing other application or the whole system stability. The VTV enable GCC is only expose to userspace after the configuration of userspace using "environment-modules".
- All dependencies of application can be expose to userspace using custom folder hierarchy. Application folder hirarchy is maintained independently to the other application hierarchy that allow to ensure application reliability and maintainability without influencing other application or the whole system.
- The userspace configuration (using  environment-modules) is commonly dedicated to a unique high computing application a time. These dedicated environments exclude many issues related to interoperate with libraries built without `-fvtable-verify`.
- Performance is important in an HPC environment, and performance testing of an application is usually carried out in a validation environment before the application is exposed to the production user space. The figures from these benchmarks enable the manager of an application's file hierarchy to reject a version (optimizations flags can be used to reduce the impact on performance).  As you can see from Caroline Tice's articles, performance varies from benchmark to benchmark. Some applications will be rejected for performance issues, others accepted.

I understand that this type of environment can be unusual and may differ from a common system architecture (for example, a single version of a library, application and compiler maintained with the package manager), but it seems reliable to expose to userspace on-demande VTV-compatible compilers, applications and libraries in this way.

Regards,

Reply:
So this is for a specialized internal ecosystem, not a best practice for the wider open source ecosystem. So arguably off topic in this guide then.

Reply:
Multi-version computer applications on custom folder hierarchies are common ecosystems. You can find many examples using this installation architecture, for example :
* https://docs.rcc.fsu.edu/hpc/environment-modules/
* https://docs.hpc.shef.ac.uk/en/latest/stanage/software/development/gcc.html

Moreover, most of project used on this king of architecture are opensource. A simple way to visualize this architecture is to compare it with installations prefixed to custom folders when configuring compilations (./configure --prefix=/path/) for the application and its dependencies. This practice is common in open-source ecosystems.

I'm not comparing it to a specialized internal ecosystem, but that's not an argument for including the -fvtable-verify flag in this guide. I'm not the right person to allow myself to justify the inclusion of this flag in this guide.